File: src/Core/OntologySourceReference.fs
namespace ARCtrl

open ARCtrl.Helper
open Fable.Core

[<AttachMembers>]
type OntologySourceReference(?description, ?file, ?name, ?version, ?comments) =
    
    let mutable _description : string option = description
    let mutable _file : string option = file
    let mutable _name : string option = name
    let mutable _version : string option = version
    let mutable _comments : ResizeArray<Comment> = Option.defaultValue (ResizeArray()) comments

    member this.Description
        with get() = _description
        and set(description) = _description <- description

    member this.File
        with get() = _file
        and set(file) = _file <- file

    member this.Name
        with get() = _name
        and set(name) = _name <- name

    member this.Version
        with get() = _version
        and set(version) = _version <- version

    member this.Comments
        with get() = _comments
        and set(comments) = _comments <- comments


    static member make description file name version comments = 
        OntologySourceReference(?description=description, ?file=file, ?name=name, ?version=version, comments=comments)

    static member create(?description,?file,?name,?version,?comments) : OntologySourceReference =
        OntologySourceReference.make description file name version (Option.defaultValue (ResizeArray()) comments)

    static member empty =
        OntologySourceReference.create()

    member this.Copy() =
        let nextComments = this.Comments |> ResizeArray.map (fun c -> c.Copy())
        OntologySourceReference.make this.Description this.File this.Name this.Version nextComments

    override this.GetHashCode() : int = 
        [|
            HashCodes.boxHashOption this.Description
            HashCodes.boxHashOption this.File
            HashCodes.boxHashOption this.Name
            HashCodes.boxHashOption this.Version
            HashCodes.boxHashSeq this.Comments
        |]
        |> HashCodes.boxHashArray 
        |> fun x -> x :?> int

    override this.Equals(obj) =
        match obj with
        | :? OntologySourceReference as t -> this.GetHashCode() = t.GetHashCode()
        | _ -> false

File: js/Core/OntologySourceReference.js
import { defaultArg, unwrap } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { ResizeArray_map } from "./Helper/Collections.js";
import { boxHashSeq, boxHashOption, boxHashArray } from "./Helper/HashCodes.js";
import { safeHash } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { class_type } from "../fable_modules/fable-library-js.4.22.0/Reflection.js";

export class OntologySourceReference {
    constructor(description, file, name, version, comments) {
        this._description = description;
        this._file = file;
        this._name = name;
        this._version = version;
        this._comments = defaultArg(comments, []);
    }
    get Description() {
        const this$ = this;
        return unwrap(this$._description);
    }
    set Description(description) {
        const this$ = this;
        this$._description = description;
    }
    get File() {
        const this$ = this;
        return unwrap(this$._file);
    }
    set File(file) {
        const this$ = this;
        this$._file = file;
    }
    get Name() {
        const this$ = this;
        return unwrap(this$._name);
    }
    set Name(name) {
        const this$ = this;
        this$._name = name;
    }
    get Version() {
        const this$ = this;
        return unwrap(this$._version);
    }
    set Version(version) {
        const this$ = this;
        this$._version = version;
    }
    get Comments() {
        const this$ = this;
        return this$._comments;
    }
    set Comments(comments) {
        const this$ = this;
        this$._comments = comments;
    }
    static make(description, file, name, version, comments) {
        return new OntologySourceReference(unwrap(description), unwrap(file), unwrap(name), unwrap(version), comments);
    }
    static create(description, file, name, version, comments) {
        const comments_1 = defaultArg(comments, []);
        return OntologySourceReference.make(description, file, name, version, comments_1);
    }
    static get empty() {
        return OntologySourceReference.create();
    }
    Copy() {
        const this$ = this;
        const nextComments = ResizeArray_map((c) => c.Copy(), this$.Comments);
        const description = this$.Description;
        const file = this$.File;
        const name = this$.Name;
        const version = this$.Version;
        return OntologySourceReference.make(description, file, name, version, nextComments);
    }
    GetHashCode() {
        const this$ = this;
        return boxHashArray([boxHashOption(this$.Description), boxHashOption(this$.File), boxHashOption(this$.Name), boxHashOption(this$.Version), boxHashSeq(this$.Comments)]) | 0;
    }
    Equals(obj) {
        let t;
        const this$ = this;
        return (obj instanceof OntologySourceReference) && ((t = obj, safeHash(this$) === safeHash(t)));
    }
}

export function OntologySourceReference_$reflection() {
    return class_type("ARCtrl.OntologySourceReference", undefined, OntologySourceReference);
}

export function OntologySourceReference_$ctor_7C9A7CF8(description, file, name, version, comments) {
    return new OntologySourceReference(description, file, name, version, comments);
}

